build_tools/github_actions/get_latest_green.sh

```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Finds the most recent ancestor (first parents only) of the given commit
+# (default HEAD) for which the ci workflow passed. There are GitHub actions that
+# allegedly do this, but they mostly seem to be broken, unlicensed, or both.
+# This approach iterates through commits in the history and makes a separate
+# call for each. The alternative is collecting all the successful runs and then
+# sorting them. Empirically, the latter was much slower. Fetching them already
+# sorted does not appear to be an option as GitHub does not document any known
+# sort order. Fetching for multiple specific commits at once is similarly not an
+# option.
+
+set -euo pipefail
+
+commitish="${1:-HEAD}"
+
+function get_latest_green() {
+  local -a query_params=(
+    branch=main
+    event=push
+    status=success
+    exclude_pull_requests=true
+  )
+
+  while read commit; do
+    local -a local_query_params=("${query_params[@]}" "head_sha=${commit}")
+    # String join on comma
+    query_string="$(IFS="&" ; echo "${local_query_params[*]}")"
+
+    successful_runs="$(gh api --jq '.total_count' "/repos/iree-org/iree/actions/workflows/ci.yml/runs?${query_string}")"
+
+    if (( successful_runs!=0 )); then
+      echo "${commit}"
+      return 0
+    fi
```

Comment:
I was a bit confused about why this was checking successful runs. Since this is only checking ci.yml, I guess it's looking for at least one run (including reruns) that passed? If we checked `failed_runs == 0`, we'd miss successful reruns?

Maybe we also want to check https://github.com/iree-org/iree/blob/main/.github/workflows/samples.yml and https://github.com/iree-org/iree/blob/main/.github/workflows/run_shark_tank.yml too? Or at least elaborate on "for which the ci workflow passed" above to specifically mention the workflow that it checks, and maybe rename from "get_latest_green" to "get_latest_commit_passing_ci"

Reply:
`successful_runs` is just a variable I set above. I guess it could be called "successful_run_count". It just looks at the `total_count` parameter returned by the above query, which filters for only successful runs. We also want to exclude anything for which the CI is still running. It's discussed in the attached issue, but GitHub doesn't give us very helpful results if we query for the "status" of a commit because it returns all the suites, including ones for workflows that don't run on main (e.g. the CLA check). That's one of the reasons the action we're currently using is broken. We could add additional workflows if desired, though I'm not sure whether one that's only on a cron is a good match

Reply:
An alternative to doing it by workflow would be doing it by check run name. For some reason, these don't include the name of their parent workflow, so we'd have job name conflict. Theoretically, this would allow us to see the status of things other than GitHub actions, but as it turns out, none of them use the new "check suites" and "check runs" API, which GitHub has helpfully made incompatible with the old "status" API.

Reply:
If this were checking multiple workflows, `successful_runs!=0` would read as "if _any_ run succeeded", when we really mean "if _all_ runs succeeded" or "if _no_ runs _failed_"

I'm on the fence about checking cron workflows. We probably don't want to create a stable release from a commit that failed community builds or complex samples (assuming those tests are reliable enough).

Reply:
Ah I see your confusion. Yeah this query is workflow specific, so indeed we want any successful run because of reruns

Reply:
If we were checking multiple workflows it would have to be multiple separate queries and we'd check that each succeeded.

This workflow is for unstable releases. For stable releases, I think we should be running them through a battery of tests prior to promotion. If there are things that we want to gate unstable release but not run every commit, I think we should be coordinating the run of them with the release workflow rather than hoping they've run on a recent enough commit. That all seems out of scope for this though as the current state is that it's just selecting the most recent commit where any check suite reported success, which is strictly worse than what we have now.

Reply:
SGTM. How do you feel about renaming this script to something like `get_latest_commit_passing_ci.sh` and/or adding some comments around the `if (( successful_runs!=0 ));` code?

Reply:
I added a variable and a for loop over workflows, with only ci.yml in the list as of now.